core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestPlanNodeSearcher.java

```diff
@@ -0,0 +1,109 @@
+import io.trino.sql.planner.PlanNodeIdAllocator;
+import io.trino.sql.planner.iterative.rule.test.PlanBuilder;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.PlanNodeId;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.planner.plan.ValuesNode;
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.sql.planner.plan.JoinNode.Type.INNER;
+import static io.trino.testing.assertions.Assert.assertEquals;
+
+public class TestPlanNodeSearcher
+{
+    private static final PlanBuilder BUILDER = new PlanBuilder(new PlanNodeIdAllocator(), new AbstractMockMetadata() {});
+
+    @Test
+    public void testFindAll()
+    {
+        int size = 10;
+        ProjectNode root = BUILDER.project(Assignments.of(), BUILDER.values());
+        for (int i = 1; i < size; i++) {
+            root = BUILDER.project(Assignments.of(), root);
+        }
+
+        List<PlanNodeId> rootToBottomIds = new ArrayList<>();
+        PlanNode node = root;
+        while (node instanceof ProjectNode) {
+            rootToBottomIds.add(node.getId());
+            node = ((ProjectNode) node).getSource();
+        }
+
+        List<PlanNodeId> findAllResult = PlanNodeSearcher.searchFrom(root)
+                .where(ProjectNode.class::isInstance)
+                .findAll()
+                .stream()
+                .map(PlanNode::getId)
+                .collect(toImmutableList());
+
+        assertEquals(rootToBottomIds, findAllResult);
+    }
+
+    @Test
+    public void testFindAllMultipleSources()
+    {
+        List<JoinNode> joins = new ArrayList<>();
+        for (int i = 0; i < 4; i++) {
+            joins.add(BUILDER.join(INNER, BUILDER.values(), BUILDER.values()));
+        }
+        JoinNode leftSource = BUILDER.join(INNER, joins.get(0), joins.get(1));
+        JoinNode rightSource = BUILDER.join(INNER, joins.get(2), joins.get(3));
+        JoinNode root = BUILDER.join(INNER, leftSource, rightSource);
+
+        ImmutableList.Builder<PlanNodeId> idsInPreOrder = ImmutableList.builder();
+        joinNodePreorder(root, idsInPreOrder);
+
+        List<PlanNodeId> findAllResult = PlanNodeSearcher.searchFrom(root)
+                .where(JoinNode.class::isInstance)
+                .findAll()
+                .stream()
+                .map(PlanNode::getId)
+                .collect(toImmutableList());
+
+        assertEquals(idsInPreOrder.build(), findAllResult);
+    }
+
+    /**
+     * This method adds PlanNodeIds of JoinNodes to the builder in pre-order.
+     * The plan tree must contain only JoinNodes and ValuesNodes.
+     */
+    private static void joinNodePreorder(PlanNode root, ImmutableList.Builder<PlanNodeId> builder)
+    {
+        if (root instanceof ValuesNode) {
+            return;
```

Comment:
This should add the id of the ValuesNode to `builder`

Reply:
It is supposed to collect the ids of JoinNodes only. The result is then compared with the result of `findAll()` that also collects joins only.
I wanted the simplest possible example with multiple branches in the plan. It is complementary to the preceding test method, which tests `findAll()` on a linear sequence.